### PR TITLE
Ornaments

### PR DESCRIPTION
**Pull request by Shadowmech88**
tree_with_topper ornaments_on_tree

Adds ornaments. Boxes of ornaments can be printed from the autolathe for one glass sheet per box.
Ornaments can be colored via application of crayons and pens.
Ornaments can be hung on any flora, such as potted plants or large trees, and will show up on the plant's sprite wherever you click on it.
Tree toppers are ornaments of which only one can be hung per flora. Each regular ornament box has a 10% chance to contain a star topper.

:cl:
 * rscadd: Adds ornaments. Boxes of ornaments can be printed from the autolathe. Each box of normal ornaments has a 10% chance to contain a star tree topper. Ornaments can be colored via application of crayons and pens. Ornaments can be hung on flora such as trees and potted plants, and will show up on the plant's sprite wherever you click on it.

### CODE REVIEW

File: code/game/objects/items/ornaments.dm
Comment (PJB3005) [inline code/game/objects/items/ornaments.dm]: I'm not keen on this.

I'd much prefer it if the object's name and such simply changed rather than spawning new objects.

File: code/game/objects/items/ornaments.dm
Comment (PJB3005) [inline code/game/objects/items/ornaments.dm]: This is quite the hack.

**Comment by 9600bauds** [comment]
dreamseeker_2016-12-22_18-23-25

**Comment by Shadowmech88** [comment]
My resolution isn't high enough for anything over 32x32 to fit on the screen and 32x32 itself is too small.

**Review by PJB3005 — CHANGES_REQUESTED** [review]

File: code/game/objects/items/ornaments.dm
Comment (Shadowmech88) [reply code/game/objects/items/ornaments.dm]: Yeah but what if mappers want to spawn a particular color of ornament for festive mapping?
Isn't it better for them to be able to spawn specific types than spawning a general ornament item and varediting it?

Comment (Shadowmech88) [reply code/game/objects/items/ornaments.dm]: Unless you mean keep the children but just don't use them for this, which seems a little strange to me, but I wouldn't mind.

File: code/game/objects/items/ornaments.dm
Comment (Shadowmech88) [reply code/game/objects/items/ornaments.dm]: I'll change it if you have a better way.

Comment (PJB3005) [reply code/game/objects/items/ornaments.dm]: I think not calling `..()` on the subtype would be enough, can't think of any code that uses base item `attackby`.

File: code/game/objects/items/ornaments.dm
Comment (PJB3005) [reply code/game/objects/items/ornaments.dm]: Eh it's fine I guess.

File: code/game/objects/items/ornaments.dm
Comment (Shadowmech88) [reply code/game/objects/items/ornaments.dm]: The pAI code does now, though I doubt pAI-integrated ornaments would ever be a thing.

But it's possible that someone else may in the future put code into `/obj/item/attackby()` or `/obj/attackby()`.

**Comment by ghost** [comment]
Could you wait until #12784 is merged to merge this so I don't have to fix fucken dmi conflicts again aaaaaa

File: code/game/objects/items/ornaments.dm
Comment (PJB3005) [reply code/game/objects/items/ornaments.dm]: Hrm, you're right.

Not sure then. Maybe an OOP list on the type so you can remove the copy paste entirely?

Comment (Shadowmech88) [reply code/game/objects/items/ornaments.dm]: Is there a way to create a list with custom indexes without manually adding elements one by one?

Comment (PJB3005) [reply code/game/objects/items/ornaments.dm]: You mean with numeric indices or assoc indices?

For numeric sadly not, but for assoc you can just do `list("x" = 10, "c" = 20)`

**Comment by ihadtoregisterforthis** [comment]
Use x2 mode for the screenshot only, looks better

File: code/game/objects/items/ornaments.dm
Comment (Shadowmech88) [reply code/game/objects/items/ornaments.dm]: Cool, then I can make a list var associating ornament types with the colors and just override the list for `/ornament/teardrop`.

**Comment by PJB3005** [comment]
No.

Use BYOND's built in screenshotting. It's at perfect resolution and objectively the best.

**Comment by Shadowmech88** [comment]
* Ornaments can now be hung on any flora, this includes large trees as well as potted plants.

**Comment by Shadowmech88** [comment]
* Added tree toppers, which can only be hung once per flora. Each box of normal ornaments has a 10% chance to contain a star topper.

**Review by PJB3005 — APPROVED** [review]

**Pull request merged**
